### ISSUE TITLE
Search API DB + Patch breaks MSSQL and POSTGRES

### ISSUE THREAD

**Issue by mxlav** [wet-boew/wet-boew-drupal#379]
Our organisation is going with postgreSQL. I am attempting to install the latest wetkit install and it is taking over an hour just to install the modules.

Is anyone using postgreSQL? what kind of tweaks have you made to the postgreSQL config?

**Comment by sylus** [comment]
An hour is a crazy long time something is for sure wrong with your server config. That being said postgresql should work, might add it as a test for our continuous integration Travis build.

**Comment by sylus** [comment]
Did you still need a call today around 12:00?

**Comment by sylus** [comment]
A browser based install should only take about 10-15 mins, where a drush si based install should only take about 2-3 mins.

**Comment by mxlav** [comment]
ya, the install i'm doing was brought from home :-)

**Comment by mxlav** [comment]
i had time to run a mysql setup just for the fun of it. took a few minutes.. postgreSQL setup still going..crazy slow! This is bound to end badly!

**Comment by mxlav** [comment]
Just died at around 91%, module 116

 SQLSTATE[42601]: Syntax error: 7 ERROR: syntax error at or near "CONVERT" LINE 1: ALTER TABLE search_api_db_database_node_index_nid CONVERT TO... ^

**Comment by sylus** [comment]
Yeah there is likely some tests that need to be run on a postgres setup. Right now only validated run on mysql. Will look into this. ^_^

**Comment by sylus** [comment]
Found the problem it is to do with one of the patches we are using: http://drupal.org/files/search_api_db_utf8-1144620-38.patch.

The "convert to" is only valid for mysql. If we fixed this statement all problems should be fixed just not sure how to go about this yet.

**Comment by sylus** [comment]
Filed an updated issue here: http://drupal.org/node/1144620#comment-6459444

**Comment by sylus** [comment]
Hey @RobJohnston I saw your post on the issue queue. Was thinking about trying your idea about using collate on the Select query. I'll try out MySQL are you able to try and see if it works for MSSQL?

I'd like to get this bug fixed as think it is the only blocker for our distro working in these other databases which is key.

Marking this issue as Critical.

**Comment by RobJohnston** [comment]
Yes, I'll be here for another hour and I can try postgreSQL when I get home.

**Comment by sylus** [comment]
Hmm after looking at this might be tricky to implement but will ask around some smarter people then I about this problem.

Another temporary solution we could do is detect the db environment and put the conditional logic in a switch statement based on the database. 

So we could keep our existing patch for mysql but alter the column to nvarchar for mssql and then see how POSTGRES expects this. Not too sure about this either lol.

**Comment by RobJohnston** [comment]
Looking at the installation instructions for the Drupal 7 SQL Server driver, it states:

"The database must be created using a unicode case-insensitive collation (for example the default SQL_Latin1_General_CP1_CI_AS"

So I don't think the issue at Drupal would even apply to MSSQL, unless I'm missing the point (because AS = accent sensitive).  

I hadn't seen the switch statement based on database before, but I've seen modules that I wish had it!

Just read a bit about postgres collation and I think that even depends on the OS more than anything else... doesn't look pretty.

**Comment by sylus** [comment]
Hmm we also could as an aside just disable search_api_db if it isn't being installed on a mysql db? Another option is if while installing the distro sees an apache solr server it will completely bypass search api db.

In the end might just have to disable search api db for certain environments. Makes the search less awesome out of the box but small trade off.

**Comment by mxlav** [comment]
I vote for the quick resolution (removing the broken module) in order to allow the install to complete.

**Comment by RobJohnston** [comment]
... or just remove this one patch instead of the whole module, which (I think) would only have the effect of no case-sensitivity on searches of mySQL databases?

**Comment by sylus** [comment]
Unfortunately removing the patch breaks the install as Search API would then bomb on trying to index content with accents. I may just leave the enabling on this module as something optional.

**Comment by joejoseph00** [comment]
Indexing bilingual content is more important than postgres support imho.  We've reported the bug, lets let the upstream take care of it, is there really anyone using Postgres?

**Comment by mxlav** [comment]
cbsa uses postgres

**Comment by joejoseph00** [comment]
CBSA could put a bounty on that bug, speed up the process of getting it fixed in the upstream

**Comment by RobJohnston** [comment]
Been looking a little more at how SQL could be applied only to a certain database.  How does something like this sound?

<pre>
global $databases;
$driver = $databases['default']['default']['driver'];
$database = $databases['default']['default']['database'];
switch ($driver) {
    case 'mysql':
        db_query("ALTER TABLE {$table} CONVERT TO CHARACTER SET 'utf8' COLLATE 'utf8_bin'");
        break;
    case 'pgsql':
        //Postgres installation requires UTF8.
        break;
    case 'sqlsrv':
        //MS SQL Server should have been set-up with a unicode accent sensitive collation.
        //db_query("ALTER DATABASE {$database} COLLATE SQL_Latin1_General_CP1_CI_AS");
        break;
}
</pre>

**Comment by sylus** [comment]
Hey @RobJohnston thanks for delving further into this I think this would work out great unless anyone else disagrees. For the sqlsrv case though should the db_query not be used as it will be assumed it was setup properly?

Let me know and i'll commit the code shortly and test the build :)

**Comment by RobJohnston** [comment]
Hi @sylus.  I think that because the installation instructions for MSSQL state that it should be accent-sensitive then the query shouldn't be used by default (I left in commented out in the example).  Maybe something like the checkEncoding function in the /includes/database/pgsql/install.inc file should be added in this case.  

Also, I see there is a sqlite database given in the default install... don't know what to do for that one, if anything.  Maybe just add an empty "case" statement for now.

**Comment by joejoseph00** [comment]
Hi @RobJohnston , can you clarify MSSQL?  do you mean MySQL?  was this just a type-o?  Not many people using MSSQL in this space.

**Comment by RobJohnston** [comment]
Hi @joejoseph00.  Not a spelling mistake... meaning of MSSQL was Microsoft SQL Server.  Maybe I'll say sqlsrv from now on, as that is what's returned from Drupal when you query the database driver.

**Comment by joejoseph00** [comment]
If we're adding MSSQL, we might as well add oracle as well 

++
case 'oracle':
        //Oracle character set alter  , {$database} could be an optional parameter, default is the current db
        //http://oracle.ittoolbox.com/documents/changing-the-character-set-of-an-oracle-database-18618
        db_query("ALTER DATABASE {$database} CHARACTER SET utf8_unicode_ci");
        break;

//according to the url above, they recommend restarting the db, while we do have oracle here, I could in theory test this to see if a restart is just "recommended" or actually "required", as there is a difference.

**Comment by RobJohnston** [comment]
And if anyone is wondering, from my quick reading, it looks like the SQLite ALTER statement cannot change the collation, so a new table would have to be created with the correct collations on certain columns, the data moved, the old table dropped, and then new table renamed to the old table's name.  SQLite is case-insenstive by default but only for the ASCII range of characters.  That's a more complex case!

**Comment by RobJohnston** [comment]
Thought you'd have it by now @sylus.  I've never done a patch against Drupal before, but how's this to start you off:

<pre>
diff --git a/search_api_db.install b/search_api_db.install
index d073dda..0ef5226 100644
--- a/search_api_db.install
+++ b/search_api_db.install
@@ -65,3 +65,47 @@ function search_api_db_update_7101() {
       ->execute();
   }
 }
+
+/**
+ * Implements hook_update_N().
+ *
+ * Solves index problems with umlauts, accented characters, etc. issue http://drupal.org/node/1144620.
+ */
+function search_api_db_update_7102() {
+  global $databases;
+  $search_api_tables = db_query("show tables LIKE '%search_api_db%';")->fetchCol();
+  
+  $database = $databases['default']['default'];
+  $db_driver = $database['driver'];
+  $db_name = $database['database'];
+
+  switch ($db_driver) {
+    case 'mysql':
+      foreach ($search_api_tables as $table) {
+        db_query("ALTER TABLE {$table} CONVERT TO CHARACTER SET 'utf8' COLLATE 'utf8_bin'")
+        ->execute();
+      }
+      break;
+    case 'oracle':
+      //TODO:  Check the encoding and change if required.
+      //db_query("ALTER DATABASE {$database} CHARACTER SET utf8_unicode_ci");
+      break;
+
+    case 'pgsql':
+      //PostgreSQL requires UTF8 from checkEncoding(), so no change necessary.
+      break;
+  
+    case 'sqlite':
+      //TODO:  The SQLite ALTER statement cannot change the collation, so a new 
+      //table would have to be created with the correct collations on certain 
+      //columns, the data moved, the old table dropped, and then new table 
+      //renamed to the old table's name.
+      break;
+
+    case 'sqlsrv':
+      //TODO:  Check the encoding and change if required.
+      //MS SQL Server instructions state collation as unicode accent-Sensitive.
+      //db_query("ALTER DATABASE {$db_name} COLLATE SQL_Latin1_General_CP1_CI_AS");
+      break;
+  }
+}
\ No newline at end of file
diff --git a/service.inc b/service.inc
index 6b6e7f6..f12a0ed 100644
--- a/service.inc
+++ b/service.inc
@@ -192,6 +192,24 @@ class SearchApiDbService extends SearchApiAbstractService {
 
     $set = $this->setDb();
     db_create_table($name, $table);
+    
+    //Some DBMS's will need a collation set.
+    global $databases;
+    $database = $databases['default']['default'];
+    $db_driver = $database['driver'];
+    $db_name = $database['database'];
+    
+    switch ($db_driver) {
+      case 'mysql':
+        db_query("ALTER TABLE {$name} CONVERT TO CHARACTER SET 'utf8' COLLATE 'utf8_bin'")->execute();
+        break;
+      case 'oracle':
+      case 'pgsql':
+      case 'sqlite':
+      case 'sqlsrv':
+        break;                    
+    }
+    
     if ($set) {
       $this->resetDb();
     }
</pre>

**Comment by RobJohnston** [comment]
The patch above got me through both sqlsrv and postgres installations.  You were correct @mxlav, the graphical postgres installation is as slow as molasses!  It took all afternoon.  Seemed to start off fine, but then got slower and slower.  My first though is to increase the memory limit in the php.ini file.  My second thought is to try it with the Drush command and replace "mysql://" with "pgsql://"... see if that works.

**Comment by mxlav** [comment]
Very cool..I'll give try later. I still don't have a working copy of GIT or drush at work, so, I'll give it a try at home.

**Comment by RobJohnston** [comment]
Did you have any luck yet @mxlav?  I maybe should have said that the patch was applied against the module that is downloadable from Drupal and it wasn't a patch against the patch.

**Comment by mxlav** [comment]
Hey @RobJohnston, I just finished running the install on postgres, took 1:50 minutes, but it did complete!

**Comment by sylus** [comment]
Looking at this patch now sorry for delay ^_^

**Comment by RobJohnston** [comment]
I added a little something at http://drupal.org/node/1144620#comment-6558840 that hopefully other can add to or comment upon.

**Comment by sylus** [comment]
Waiting on Travis to test the patch

**Comment by sylus** [comment]
Successfully passed Travis so looks good to merge. Thanks guys! Will close this issue for now but feel free to reopen. As for slowness of PostGres may have to make another issue.
